testsuite/libffi.call/callback3.c

```diff
@@ -0,0 +1,113 @@
+/* Area:	ffi_call
+   Purpose:	Check structures with array and callback.
+   Limitations:	none.
+   PR:		none.
+   Originator:	David Tenty <daltenty@ibm.com>  */
+
+/* { dg-do run } */
+#include "ffitest.h"
+
+int i=5;
+
+void callback() { i++; }
+
+
+typedef struct
+{
+  struct { unsigned char c; double d; } s1;
+  double s[2];
+  unsigned char c2;
```

Comment:
Can you explain what is interesting about this case?

A case that is more recursive would be more interesting. Ditto for having a case where the struct-alignment-when-nested must not be confused with the struct-alignment-for-self-layout.

```suggestion
  unsigned char c1;
  struct { double d; unsigned char c; } s[2];
  unsigned char c2;
```

Reply:
> Can you explain what is interesting about this case?

Simply that it exercises the return path in the layout computation when no nested_first_member is found. This is supposed to be the less interesting case that exercises that path.

> Ditto for having a case where the struct-alignment-when-nested must not be confused with the struct-alignment-for-self-layout

That's already the case in test case 1, where the struct we use in modeling the array has a different layout on its own.

I can add the additional more recursive case, but I prefer to add it as it's own case.

Reply:
> That's already the case in test case 1, where the struct we use in modeling the array has a different layout on its own.

It has a different alignment on its own, but it has no tail padding. This case has tail padding.

Reply:
I've added the case which has tail padding. The tail padding case does indeed have some additional issues, so I've made some changes to how we calculate the tail padding.